[Subtitles] Implemented new subtitle features

CastagnaIT opened the pull request:
## Description
<!--- Provide a general summary of your change in the Pull Request title above -->
<!--- Describe your change in detail here. -->
These changes implement new subtitles features and enhance existing ones

**New features:**
- Add new subtitle settings: 
  - Border size
  - Border color
  - Background type (list values: None/Shadow/Box/Square box)
  - Background color + opacity
  - Shadow color + opacity
  - Blur
- Add text strikethrough style support to SubRip / MicroDVD / SAMI / MPL2
- Add text alignment for CC subtitles, to improve the readability on live videos (in particular with karaoke text)
- Add related new GUI dialog window, DialogColorPicker
- Add related new GUI setting button, colorbutton
- Add advanced setting `subtitleverticalmargin`, allows to change the default hardcoded vertical position mergin of the text, in order to be applied to all type of subtitle positions.


**Changes:**
- There is no more hardcoded colors lists for subtitles settings, replaced by the color dialog window
- The subtitle color setting has been replaced to be used with the new GUI controls
- Changed default subtitle font style from Bold to Normal
- In the screen calibration, now the subtitle bar can be moved in to full screen height
- Add support to `colordiffuse` tag to handle `$INFO[...]` values (no changes to existings controls except for "Image", read bug fixes)
- Increased of 10px the default vertical margin (compared to #20010) a bit better


**Bug fixes:**
- For "Image" control, the tag `colordiffuse` can now be used to set the color with `$INFO[...]` to reflect e.g. a ListItem property in a list/panel (before if an Image control was inside a list/panel, to all Image controls were wrongly set to the color of the property of the selected ListItem, this made it impossible to relate the control xml property to a ListItem property).


**Info about the new GUI controls:**
The GUI Colorbutton setting and GUI DialogColorPicker has been added to Kodi skins purpose only, so no Python add-ons, at least for now.

The GUI DialogColorPicker, open by default a list of colors from a XML file, that is placed in:
system/dialogcolors.xml

The palette of this file has been taken from Crayola that have a good mix of colors with nice names,
is not included exactly all colors but most of them, except for greyscale where i have add a scale separately.

The colors are displayed in this order: hue --> saturation --> lightness
is a simple sort order algorithm, complex sort order algorithms are out of my skill and probably don't matter much.


**Info about the Background type "Box":**
The "Box" backgroud effect has some limitations due to LibAss rendering,
the font border size and border color cannot be set
~it is also possible to notice that in situations where there are several lines of subtitles, one box overlaps with the other:~ (fixed)

**Info about text alignment for CC subtitles:**
Has been set by default Left text alignment,
the effect displayed will be this
![CCSubAmin](https://user-images.githubusercontent.com/3257156/134720195-fa4115fe-6eaf-4980-b524-7b9e72d6c085.gif)
you notice that the text is also slightly centered on the screen to allow easier reading,
a similar text centering happens with some television broadcasters.
Centering is calculated 1/10 of the screen frame.


## Motivation and context
<!--- Why is this change required? What problem does it solve? -->
<!--- If it fixes an open issue, please link to the issue here. -->
After the subtitle rendering rework with Libass #20010
thanks to LibAss rendering this PR implement new features and solve some forum requests:
https://forum.kodi.tv/showthread.php?tid=364471
https://forum.kodi.tv/showthread.php?tid=337741
https://forum.kodi.tv/showthread.php?tid=344515
https://forum.kodi.tv/showthread.php?tid=363771
https://forum.kodi.tv/showthread.php?tid=364389
https://forum.kodi.tv/showthread.php?tid=354578
https://forum.kodi.tv/showthread.php?tid=234896
https://forum.kodi.tv/showthread.php?tid=302547
https://forum.kodi.tv/showthread.php?tid=338678


## How has this been tested?
<!--- Please describe in detail how you tested your change. -->
<!--- Include details of your testing environment, and the tests you ran to -->
<!--- see how your change affects other areas of the code, etc. -->
By using CC Subtitles with karaoke effects
By using SRT file and playing with subtitles settings


## Screenshots (if appropriate):
**New settings and GUI controls**
Estuary skin
![SS01MenuEstuary](https://user-images.githubusercontent.com/3257156/134874491-b4d4c84f-5aba-4656-b017-88d3725f78fd.jpg)
![SS01EstuaryPicker](https://user-images.githubusercontent.com/3257156/134874498-c0ade285-8762-48da-8671-833d57d408b7.jpg)

Estouchy skin
![SS02MenuEstuchy](https://user-images.githubusercontent.com/3257156/134874536-3cd6e8cb-a7a8-45bc-beb2-d4cce7442256.jpg)
![SS02EstuchyPicker](https://user-images.githubusercontent.com/3257156/134874550-f882ff00-0f76-4b12-9a04-43e9197bf697.jpg)


Example of customized subtitle style
![SS5](https://user-images.githubusercontent.com/3257156/134720318-913de3c7-e85c-4ac7-a165-b3338226560b.jpg)

## Types of change
<!--- What type of change does your code introduce? Put an `x` in all the boxes that apply like this: [X] -->
- [x] **Bug fix** (non-breaking change which fixes an issue)
- [ ] **Clean up** (non-breaking change which removes non-working, unmaintained functionality)
- [x] **Improvement** (non-breaking change which improves existing functionality)
- [x] **New feature** (non-breaking change which adds functionality)
- [ ] **Breaking change** (fix or feature that will cause existing functionality to change)
- [ ] **Cosmetic change** (non-breaking change that doesn't touch code)
- [ ] **None of the above** (please explain below)

## Checklist:
<!--- Go over all the following points, and put an `X` in all the boxes that apply like this: [X] -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [x] My code follows the **[Code Guidelines](https://github.com/xbmc/xbmc/blob/master/docs/CODE_GUIDELINES.md)** of this project 
- [x] My change requires a change to the documentation, either Doxygen or wiki
- [ ] I have updated the documentation accordingly
- [x] I have read the **[Contributing](https://github.com/xbmc/xbmc/blob/master/docs/CONTRIBUTING.md)** document
- [ ] I have added tests to cover my change
- [ ] All new and existing tests passed

CastagnaIT committed:
[Subtitles] Add strikethrough text style support

CastagnaIT committed:
[Subtitles] Allow subtitle bar to move in full screen height

CastagnaIT committed:
[Subtitles] Add advanced setting subtitleverticalmargin

ksooo commented on xbmc/dialogs/GUIDialogSelectColor.h:
```diff
@@ -0,0 +1,53 @@
+
+  CGUIViewControl m_viewControl;
+  CFileItemList* m_vecList;
+  bool m_focusToButton{};
```
was the intention to initialize the bool member? you don't default-init the other members...

CastagnaIT replied:
changed

ksooo commented on xbmc/dialogs/GUIDialogSelectColor.h:
```diff
@@ -0,0 +1,53 @@
+class CGUIDialogSelectColor : public CGUIDialogBoxBase
+{
+public:
+  CGUIDialogSelectColor(void);
```
remove `void`? here and everywhere else in method declarations/definitions.

ksooo commented on xbmc/guilib/GUIButtonControl.cpp:
```diff
@@ -312,9 +312,9 @@ void CGUIButtonControl::SetAlpha(unsigned char alpha)
 }
 
-bool CGUIButtonControl::UpdateColors()
+bool CGUIButtonControl::UpdateColors(const CGUIListItem *item)
```
asterisk next to the type, not variable name. coding convention. here and everywhere else.

ksooo commented on xbmc/settings/windows/GUIControlSettings.cpp:
```diff
@@ -366,6 +368,61 @@ void CGUIControlRadioButtonSetting::Update(bool fromControl, bool updateDisplayO
+  : CGUIControlBaseSetting(id, std::move(pSetting), localizer)
+{
+  m_pColorButton = pColorControl;
+  if (m_pColorButton == NULL)
```
nullptr, not NULL. here and everwhere else

ksooo commented on xbmc/settings/windows/GUIControlSettings.cpp:
```diff
@@ -366,6 +368,61 @@ void CGUIControlRadioButtonSetting::Update(bool fromControl, bool updateDisplayO
 
+CGUIControlColorButtonSetting::CGUIControlColorButtonSetting(CGUIColorButtonControl* pColorControl,
+                                                             int id,
+                                                             std::shared_ptr<CSetting> pSetting,
```
`const std::shared_ptr<CSetting>&`? by-ref vs. by-val param.

ksooo commented on xbmc/utils/ColorUtils.cpp:
```diff
@@ -9,34 +9,165 @@
-UTILS::Color ColorUtils::ChangeOpacity(const UTILS::Color color, const float opacity)
+using namespace UTILS;
+
+
```
remove this empty line

ksooo commented on xbmc/utils/ColorUtils.cpp:
```diff
@@ -9,34 +9,165 @@
+  colorEx.lightness = l;
+}
+
+} // namespace
```
i and many other defs close anon namespaces with `} // unnamed namespace`. just a personal thing at the end

ksooo commented on xbmc/utils/ColorUtils.cpp:
```diff
@@ -9,34 +9,165 @@
+Color COLORUTILS::ConvertHexToColor(const std::string& hexColor)
+{
+  Color value = 0;
+  sscanf(hexColor.c_str(), "%x", &value);
```
I use `std::` for all those beloved ancient clib functions nowadays

ksooo commented:
Awesome! Just commented on what I saw when scrolling quickly over the changes. Not actually did a full code review.

CastagnaIT commented:
I mostly copied pre-existing code, i will make the requested changes as soon as possible

CastagnaIT committed:
[Subtitles] Add support to set colordiffuse tag value with $INFO[] property

The tag <colordiffuse> can now be used to set the color with $INFO[...] to reflect e.g. a ListItem property within of a list/panel
This change not affect behaviour in existing controls other than Image

enen92 commented on xbmc/dialogs/GUIDialogSelectColor.cpp:
```diff
@@ -0,0 +1,239 @@
+
+void CGUIDialogSelectColor::OnInitWindow()
+{
+  // If no file path provided, load the default file
```
Also please take the cyclomatic complexity into consideration when adding new methods. At least the parsing logic can be made an independent private method

CastagnaIT replied:
i have reworked all methods in similar way of select dialog window, now there are:
AddItem
SetItems
LoadColors

enen92 commented on xbmc/dialogs/GUIDialogSelectColor.cpp:
```diff
@@ -0,0 +1,239 @@
+    }
+  }
+
+  m_viewControl.SetItems(*m_vecList);
```
Shouldn't the veclist be checked for nullptr?

CastagnaIT replied:
is not necessary because it is created when the class is initialised and destroyed only when the class is destroyed

enen92 commented on xbmc/dialogs/GUIDialogSelectColor.cpp:
```diff
@@ -0,0 +1,239 @@
+  m_viewControl.SetItems(*m_vecList);
+  m_viewControl.SetCurrentView(CONTROL_ICON_LIST);
+  
+  SET_CONTROL_LABEL(CONTROL_NUMBER_OF_ITEMS,
```
Is the label/Title of the dialog always the same? You can select colours in different contexts, we likely want the title to be configurable

CastagnaIT replied:
this is just a item counter not the window title, same counter there is in all the windows with lists

enen92 commented on xbmc/dialogs/GUIDialogSelectColor.cpp:
```diff
@@ -0,0 +1,239 @@
+
+  CGUIDialogBoxBase::OnInitWindow();
+
+  // focus one of the buttons if explicitly requested
```
Then this block of code should be guarded against something that can only be set if the codeflow has hit window initialize (maybe a null check in the color vector is enough). Log an error if it doesnt

enen92 commented on xbmc/dialogs/GUIDialogSelectColor.cpp:
```diff
@@ -0,0 +1,239 @@
+  }
+
+  // if nothing is selected, select first item
+  m_viewControl.SetSelectedItem(std::max(GetSelectedItem(), 0));
```
What if the color list is empty? I.e. if you don't even have any i
selectable item

CastagnaIT replied:
nothing happen, in this case the item selection will be skipped

enen92 commented on xbmc/guilib/GUIColorButtonControl.cpp:
```diff
@@ -0,0 +1,168 @@
+void CGUIColorButtonControl::SetImageBoxColor(std::string hexColor)
+{
+  if (hexColor.empty())
+    m_imgBoxColor = UTILS::COLOR::BLACK;
```
Out of curiosity, any reason to fill it black by default if no color is defined? I'd prefer it not to have any color (transparency) if the color is not explicitly set

enen92 commented on xbmc/guilib/GUIWindowManager.cpp:
```diff
@@ -327,6 +329,7 @@ bool CGUIWindowManager::DestroyWindows()
+    DestroyWindow(WINDOW_DIALOG_SELECT_COLOR);
     DestroyWindow(WINDOW_DIALOG_OK);
     DestroyWindow(WINDOW_DIALOG_KEYBOARD);
     DestroyWindow(WINDOW_DIALOG_KEYBOARD_TOUCH);
```
I prefer color_picker for the dialog name, but it is just personal preference

enen92 commented on xbmc/settings/AdvancedSettings.cpp:
```diff
@@ -576,6 +578,7 @@ void CAdvancedSettings::ParseSettingsFile(const std::string &file)
   pElement = pRootElement->FirstChildElement("video");
   if (pElement)
   {
+    XMLUtils::GetInt(pElement, "subtitleverticalmargin", m_videoSubtitleVerticalMargin);
```
Needed? We should avoid advanced settings whenever possible

CastagnaIT replied:
has been requested in the forum and i find this extra functionality very useful,
because it allows you to move the hardcoded vertical margin to all types of subtitle position,
if no one is against, i would prefer to keep this feature, is minimally invasive

enen92 replied:
Sure, it's okay to me. The reason I asked is because it misses documentation and as such I missed your original intention. Please describe the setting if possible near the variable declaration. It will also help filling the wiki later

enen92 commented on xbmc/settings/Settings.h:
```diff
@@ -138,14 +138,21 @@ class CSettings : public CSettingsBase, public CSettingCreator, public CSettingC
   static constexpr auto SETTING_MYVIDEOS_STACKVIDEOS = "myvideos.stackvideos";
   static constexpr auto SETTING_LOCALE_SUBTITLELANGUAGE = "locale.subtitlelanguage";
   static constexpr auto SETTING_SUBTITLES_PARSECAPTIONS = "subtitles.parsecaptions";
+  static constexpr auto SETTING_SUBTITLES_CAPTIONSALIGN = "subtitles.captionsalign";
```
In the future we likely want to move subtitle setting operations to the subtitlesettings file. At least it's one of my goals

enen92 commented on xbmc/settings/dialogs/GUIDialogSettingsBase.cpp:
```diff
@@ -59,6 +61,7 @@ CGUIDialogSettingsBase::CGUIDialogSettingsBase(int windowId, const std::string&
     m_pOriginalSpin(NULL),
     m_pOriginalSlider(NULL),
     m_pOriginalRadioButton(NULL),
+    m_pOriginalColorButton(NULL),
```
Nullptr

enen92 commented on xbmc/settings/dialogs/GUIDialogSettingsBase.cpp:
```diff
@@ -756,6 +766,17 @@ CGUIControl* CGUIDialogSettingsBase::AddSetting(const std::shared_ptr<CSetting>&
     pSettingControl.reset(new CGUIControlLabelSetting(static_cast<CGUIButtonControl*>(pControl),
                                                       iControlID, pSetting, this));
   }
+  else if (controlType == "colorbutton")
```
nullptr for new added code

enen92 commented on xbmc/utils/Color.h:
```diff
@@ -15,6 +15,15 @@ namespace UTILS
 
+  struct ColorEx
+  {
+    std::string hex;
```
Does this need to be a std::string since it only refers to an hex value. Can't the color typedef be used instead?

CastagnaIT replied:
removed

enen92 commented on xbmc/utils/Color.h:
```diff
@@ -15,6 +15,15 @@ namespace UTILS
 
   typedef uint32_t Color;
 
+  struct ColorEx
```
Any other name suggestion instead of colorex? It doesn't convey the struct intention well

CastagnaIT replied:
suggestions? perhaps `ColorInfo` is better?

enen92 replied:
ColorInfo sounds good to me

enen92 commented on xbmc/utils/ColorUtils.cpp:
```diff
@@ -9,34 +9,165 @@
+    return (a.second.hue < b.second.hue);
+}
+
+bool COLORUTILS::LoadXMLFileColors(const std::string& filePath,
```
Loading colours from an XML file isn't really something I would think as an utils method. This likely belongs elsewhere although I can't say for sure where. We have a color manager but effort has been made into getting rid of it and moving it to the skin. Subtitle colours however belong to the system scope and should be always available (i.e. they don't depend depend on GUI or skins to be present (in an headless scenario we still want users to be able to configure the subtitles color as it might be present on external players). I'd try to move this into colormanager for now, or create a ColorLoader class and invoke it wherever this is actually needed

CastagnaIT replied:
i have moved it in to colormanager try give a look

enen92 reviewed (commented):
First round of comments by looking at it on the phone without runtime testing or even play with it. Thanks for fixing the vertical spacing issue in boxed subtitles and the provided features here. It goes well in line with the way I envisioned this

enen92 commented:
@ronie can you take a look at the skin (and probably GUI) changes when you have some time?

ProfYaffle commented on addons/resource.language.en_gb/resources/strings.po:
```diff
@@ -22616,3 +22616,87 @@ msgstr ""
+#. Subtitle border color setting
+#: system/settings/settings.xml
+msgctxt "#39160"
+msgid "Border color"
```
Correct: _color_ in US-EN, _colour_ in most other dialects.

CastagnaIT replied:
ok will change in "colour" i did not know the difference

CastagnaIT commented on cmake/installdata/common/common.txt:
```diff
@@ -7,6 +7,7 @@ system/settings/*
 userdata/*
 system/addon-manifest.xml
 system/colors.xml
+system/dialogcolors.xml
```
you cannot use the same xml, the colors.xml is used by skins xml and require to have the color name code to be identified, they are limited in his context. The color dialog is different thing and each color have appropriate name as hint to the user

CastagnaIT replied:
or they should be merged, but with a different xml hierarchy to distinguish the two lists

enen92 replied:
Hum, don't mean to disagree but the structure of both files seem to be identical. If the only reason for having a new file is the fact the name  in the new color picker dialog is presented on top of each color I don't consider it a strict requirement. Personally, as a user, I want to pick up a color in the dialog by its visual appearance, i.e. I don't care if the color is called "cerulean". I think those names on the color buttons are not necessary just add complexity to both the user experience or skin porting itself (don't forget whenever a new window is added skins have to implement it and we have a breaking change in the skinning API). That means that we should make an effort to make Skinner's life the easier as possible

CastagnaIT replied:
the name of the colours in the file colours.xml are mandatory to allow GUI system to set the right color when the skin components set a color name instead of an hex,
The color names "may" not be necessary in the new color list i have created, but, this is the only way to give some help to all visually impaired users, having an extra Label in the window dialog xml control is certainly not a huge job for skinners it is only few lines in the xml dialog code.
Then using the existing color list is not appropriate due to CSS color code names that are not possible to change.
so solutions could be to keep two separate files, or else use a single xml file, but keep two different colour lists in it

CastagnaIT replied:
for the information i am visually impaired, and without colour names it would be a problem

thexai replied:
Some usability/design suggestions:

Maybe color name can be shown next to color rectangle in settings (to know exactly what the current color is without having to enter to change it):

![Captura de pantalla 2021-09-25 193713](https://user-images.githubusercontent.com/58434170/134781259-6033a14a-0b71-4573-b680-c9356a0ed25f.png)

Color name currently highlighted can shown on a label below instead of inside each color:

![Captura de pantalla 2021-09-25 194326](https://user-images.githubusercontent.com/58434170/134781346-f38969c9-e0d8-4bf4-bf84-01ca4018cb7d.png)

enen92 replied:
@CastagnaIT allow me some time to test and play a bit with the code to understand the role of the current colors.xml. The reason I am being "stubborn" with this point is because it always has been a common desire to let scripts provide their own fonts and colours (one of our add-on system biggest limitations). Prior effort has been made (https://github.com/xbmc/xbmc/pull/11136) to move the resolution of colours (and fonts) to the skin object. Subtitle colours are not exactly the same thing (I don't think they should belong to the skin or even the GUI). So let's think and agree on a way forward that allow us to change the design incrementally

CastagnaIT replied:
@enen92 ok for now i do not touch the xml things until you get more info about this or/and also others provide his opinions
@thexai to move the name below it is ok, instead to add the name in the settings list, I wouldn't know how to deal with it, because we cannot store the color name on same setting, this means that every time we should load the xml colors to be allow to find the name for each color setting, would slow down the loading of the settings window

thexai replied:
Then maybe shown hex value in settings `#RRGGBB` (or the same way it is saved). Also a additional label with hex value can be shown in color picker.

jjd-uk replied:
I'd agree with @thexai regarding displaying the label over the colour, Due to the length of some of the colour names this results in the label needing to scroll.

![image](https://user-images.githubusercontent.com/5781142/134803213-b38ed323-662d-4a8f-9505-bde040005ed9.png)

Doing a quick play around in the new DialogSelectColor.xml something like this maybe better:

![image](https://user-images.githubusercontent.com/5781142/134803033-8fb687ab-de6e-4d7c-a3b7-8df9d1afa45a.png)

thexai replied:
Since there is enough space, the value in hex could also be added:

Colour: colour_name - #RRGGBB

enen92 replied:
If I had to pick one option for the color button layout (settings dialog before selecting the colour) the above @thexai suggestion would be the one (instead of colour name) I'd choose. We likely want the user to add custom colours later (either RGB or html) via a new button in the colour picker dialog (requested several times). Forcing the user to provide a name for the provided colour or simply displaying "custom" is worst than having a common way to describe them all, including custom colours

CastagnaIT replied:
if a new color is manually added we display only the hex value not the color name, the user is responsible for his additions

CastagnaIT replied:
in the hex value we have ARGB so will be shown as
Colour: colour_name - #AARRGGBB
maybe there is a way to trim the first two chars with the xml code? or i keep as is?

CastagnaIT replied:
i have applied the changes based on @thexai  @jjd-uk suggestions and updated screenshots

CastagnaIT commented on xbmc/cores/VideoPlayer/DVDCodecs/Overlay/DVDOverlayCodecTX3G.cpp:
```diff
@@ -74,9 +75,9 @@ typedef struct
-  m_textColor =
-      KODI::SUBTITLES::colors[CServiceBroker::GetSettingsComponent()->GetSettings()->GetInt(
-          CSettings::SETTING_SUBTITLES_COLOR)];
+  m_textColor = UTILS::COLORUTILS::ConvertHexToColor(
```
I have kept the name of the previous class name because I thought there was a reason why this distinction was made,
but i think there is no problem use Color namespace

CastagnaIT commented on xbmc/cores/VideoPlayer/VideoRenderers/OverlayRenderer.cpp:
```diff
@@ -282,25 +283,61 @@ void CRenderer::CreateSubtitlesStyle()
+  m_overlayStyle->shadowSize = settings->GetInt(CSettings::SETTING_SUBTITLES_SHADOWSIZE);
+
+  int subTextAlign = SUBTITLE_TEXT_ALIGN_CENTER;
+  if (o->IsOverlayType(DVDOVERLAY_TYPE_TEXT))
```
so you want to add to renderOpts struct a new property like
`FontHorizontalAlign horizontalAlign = FontHorizontalAlign::LEFT;`
?

enen92 replied:
Yeah that sounds better, something like HorizontalAlignment

enen92 commented on xbmc/cores/VideoPlayer/DVDSubtitles/DVDSubtitlesLibass.cpp:
```diff
@@ -317,58 +317,96 @@ void CDVDSubtitlesLibass::ApplyStyle(style subStyle, renderOpts opts)
+        style->OutlineColour = ConvColor(subStyle.fontBorderColor, subStyle.fontOpacity);
+        style->BackColour =
+            ConvColor(subStyle.shadowColor, subStyle.shadowOpacity); // Set the shadow color
+        style->Shadow = (10.00 / 100 * subStyle.shadowSize) * scale; // Set the shadow size
```
Since we are implementing shadows, although it might be considered a feature request, blur is something modern streaming applications (like netflix) use by default. This is part of the ASS_Style type: https://github.com/libass/libass/blob/master/libass/ass_types.h#L79

CastagnaIT replied:
i have add it

enen92 commented on xbmc/guilib/guiinfo/GUIInfoColor.cpp:
```diff
@@ -17,13 +17,19 @@
 using namespace KODI::GUILIB::GUIINFO;
 
-bool CGUIInfoColor::Update()
+bool CGUIInfoColor::Update(const CGUIListItem* item /* = nullptr */)
```
Compilation failing for me in Linux due to missing includes:

```
#include "guilib/GUIListItem.h"
#include "FileItem.h"
```

CastagnaIT replied:
thanks

thexai commented on xbmc/settings/windows/GUIControlSettings.h:
```diff
@@ -173,6 +197,27 @@ class CGUIControlListSetting : public CGUIControlBaseSetting
+
+  CGUIControl* GetControl() override { return reinterpret_cast<CGUIControl*>(m_pButton); }
+  bool OnClick() override;
+  void Clear() override { m_pButton = NULL; }
```
```suggestion
  void Clear() override { m_pButton = nullptr; }
```

thexai commented on xbmc/guilib/GUIColorButtonControl.h:
```diff
@@ -0,0 +1,71 @@
+  void SetHeight(float height) override;
+  std::string GetDescription() const override;
+  void SetColorDimensions(float posX, float posY, float width, float height);
+  bool IsSelected() const { return m_bSelected; };
```
```suggestion
  bool IsSelected() const { return m_bSelected; }
```

enen92 reviewed (approved):
I am happy with this. Awesome work! :)
Would be nice if some skinner could cross check the skin changes before merging.

(commits need to be squashed and the history bissectable before merging, but don't waste time with that now)

ronie reviewed (approved):
skin changes look good to me.
i've tested both estuary & estouchy and everything works as expected.

HitcherUK commented:
Skin changes also look good to me but I assume we'll also need a xbmc.gui bump as this adds a new dialog.

@ronie or @ksooo will know the bumping procedure better than me though.

Speaking of which, skinners are going to pleased we now have a built in colour picker so big thanks for that.

CastagnaIT committed:
[Subtitles] Implemented new subtitle features

CastagnaIT commented:
i have updated Wiki advanced setting page ref. to `subtitleverticalmargin` and `assfixedworks`

CastagnaIT committed:
[GUI] bump xbmc.gui to 5.16

CastagnaIT committed:
[Skins] bump skin version and dependencies

enen92 commented:
Guess we have a few threads to mark as solved in the forum 🚀

monsoons commented:
Could you plz add this new subtitle features to OSD SubtilteSetting?
Changing HDR <-> SDR videos, I need to change 2 settings (subtitle color and border size).
I hope I can change those setting during playback quickly without opening playersettings window.

enen92 commented:
@monsoons please keep feature requests at the forum.
Regarding your comment: I also though about that and can't seem to find a way of adding those settings there without a cumbersome user experience. I think the way to go is probably to create some sort of subtitle profiles so that you can cycle them during playback. However, implementing this is still a fair bit of work.

mohd-akram commented:
Is there any chance this feature will be added to Matrix? It's very useful.

HitcherUK commented:
> Is there any chance this feature will be added to Matrix? It's very useful.

No, only bug fixes are backported not new features.

jjd-uk commented:
In addition it's a huge change so would carry too much risk to backport.

jurialmunkey commented:
Just a reminder that the new "colorbutton" control still needs to be documented by someone under the controls section of the Skinning Manual Wiki: https://kodi.wiki/view/Skinning_Manual#Controls

I've documented the built-ins for the dialog: https://kodi.wiki/view/List_of_Built_In_Controls#DialogColorPicker.xml

However, I'm locked out editing the skinning manual due to stupid Cloudflare protection so I cannot document the control itself or add a new template for it.

eartempthree commented:
Excuse me. @CastagnaIT  Thank you for your great work!
The blur adjustment is not working for ASS subtitles while bordersize is perfectly working.
Is it a bug or a limitation?

Hitcher commented:
@KarellenX the colorbutton control still hasn't been added to the wiki.

KarellenX commented:
@Hitcher I don't edit the skinning pages. I leave that for the skinners. 
Ronie used to do it, but he is no longer around.
Someone with more skinning knowledge needs to edit.
"Wiki Done" is because I added the new feature to the wiki for users, not skinning controls for skinners.

Hitcher commented:
> "Wiki Done" is because I added the new feature to the wiki for users, not skinning controls for skinners.

Ah ok, I thought you changed it for the skinning manual.
